src/NumericalAlgorithms/RootFinding/RootBracketing.hpp

```diff
@@ -0,0 +1,317 @@
+        // iteration: the new valid point and its neighbors.
+        if (level > max_level) {
+          ERROR("Too many iterations in bracket_by_contracting");
+        }
+        // Recurse.
+        return bracket_by_contracting({{x[i], xx[2 * i + 1], x[i + 1]}},
+                                      {{y[i], yy[2 * i + 1], y[i + 1]}}, f,
+                                      level + 1);
+      }
+    }
+    if (level > max_level) {
+      ERROR("Too many iterations in bracket_by_contracting");
+    }
+    // We didn't find a valid point. So recurse, using all points.
+    return bracket_by_contracting(xx, yy, f, level + 1);
+  }
+
+  // If we get here, we have found a valid point; in particular we have
+  // found the last valid point in the array.
+
+  // Find the first valid point in the array.
+  size_t first_valid_index = 0;
+  for (size_t i = 0; i < y.size(); ++i) {
+    if (y[i]) {
+      first_valid_index = i;
+      break;
+    }
+  }
+
+  // Make a new set of points that includes only the points that
+  // neighbor the boundary between valid and invalid points.
+  std::vector<double> xx;
+  std::vector<boost::optional<double>> yy;
+
+  if (first_valid_index > 0) {
+    // Check for a root between first_valid_index-1 and first_valid_index.
+    const double x_test =
+        x[first_valid_index - 1] +
+        0.5 * (x[first_valid_index] - x[first_valid_index - 1]);
+    const auto y_test = f(x_test);
+    if (y_test and y[first_valid_index].get() * y_test.get() <= 0.0) {
+      // Bracketed!
+      return std::array<double, 4>{{x_test, x[first_valid_index], y_test.get(),
+                                    y[first_valid_index].get()}};
+    } else {
+      xx.push_back(x[first_valid_index - 1]);
+      yy.push_back(y[first_valid_index - 1]);
+      xx.push_back(x_test);
+      yy.push_back(y_test);
+      xx.push_back(x[first_valid_index]);
+      yy.push_back(y[first_valid_index]);
+    }
+  }
+  if (last_valid_index < y.size() - 1) {
+    // Check for a root between last_valid_index and last_valid_index+1.
+    const double x_test = x[last_valid_index] +
+                          0.5 * (x[last_valid_index + 1] - x[last_valid_index]);
+    const auto y_test = f(x_test);
+    if (y_test and y[last_valid_index].get() * y_test.get() <= 0.0) {
+      // Bracketed!
+      return std::array<double, 4>{{x[last_valid_index], x_test,
+                                    y[last_valid_index].get(), y_test.get()}};
+    } else {
+      if (first_valid_index != last_valid_index or first_valid_index == 0) {
+        xx.push_back(x[last_valid_index]);
+        yy.push_back(y[last_valid_index]);
+      } // else we already pushed back last_valid_index (==first_valid_index).
+      xx.push_back(x_test);
+      yy.push_back(y_test);
+      xx.push_back(x[last_valid_index + 1]);
+      yy.push_back(y[last_valid_index + 1]);
+    }
+  }
+
+  if (level > max_level) {
+    ERROR("Too many iterations in bracket_by_contracting");
```

Comment:
Should this suggest that the initial guesses be refined or maybe `max_level` be increased?